lib/ddtrace/transport/traces.rb

```diff
@@ -8,25 +9,177 @@ module Traces
+        # https://github.com/DataDog/datadog-agent/blob/6.14.1/pkg/trace/api/api.go#L46
+        #
+        # We set the value to a conservative 5 MiB, in case network speed is slow.
+        DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1024 * 1024
+
+        attr_reader :encoder, :max_size
+
+        #
+        # Single traces larger than +max_size+ will be discarded.
+        #
+        # @param encoder [Datadog::Encoding::Encoder]
+        # @param max_size [String] maximum acceptable payload size
+        def initialize(encoder, max_size: DEFAULT_MAX_PAYLOAD_SIZE)
+          @encoder = encoder
+          @max_size = max_size
+        end
+
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Enumerable<Trace>] list of traces
+        # @return [Enumerable[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.reject(&:nil?)
+
+          Datadog::Chunker.chunk_by_size(encoded_traces, max_size).map do |chunk|
+            [encoder.join(chunk), chunk.size]
+          end
+        end
+
+        private
+
+        def encode_one(trace)
+          encoded = Encoder.encode_trace(encoder, trace)
+
+          if encoded.size > max_size
+            # This single trace is too large, we can't flush it
+            Datadog.logger.debug { "Dropping trace. Payload too large: '#{trace.map(&:to_hash)}'" }
+            Datadog.health_metrics.transport_trace_too_large(1, tags: ["max_size:#{max_size}"])
+
+            return nil
+          end
+
+          encoded
+        end
+      end
+
+      # Encodes traces using {Datadog::Encoding::Encoder} instances.
+      module Encoder
+        module_function
+
+        def encode_trace(encoder, trace)
+          encoder.encode(trace.map(&:to_hash))
+        end
+      end
+
+      # Sends traces based on transport API configuration.
+      #
+      # This class initializes the HTTP client, breaks down large
+      # batches of traces into smaller chunks and handles
+      # API version downgrade handshake.
+      class Transport
+        attr_reader :client, :apis, :default_api, :current_api_id
+
+        def initialize(apis, default_api)
+          @apis = apis
+          @default_api = default_api
+
+          change_api!(default_api)
+        end
+
+        def send_traces(traces)
+          encoder = current_api.encoder
+          chunker = Datadog::Transport::Traces::Chunker.new(encoder)
+
+          responses = chunker.encode_in_chunks(traces.lazy).map do |encoded_traces, trace_count|
+            request = Datadog::Transport::Traces::Request.new(encoded_traces, trace_count)
+
+            client.send_payload(request).tap do |response|
+              if downgrade?(response)
+                downgrade!
+                return send_traces(traces.lazy)
```

Comment:
What's `traces.lazy`?

When we fail, downgrade and retry, are we re-encoding already encoded output, or doing chunking again? Unless I'm misunderstanding something, I would expect this probably to be `send_payload` no?

Reply:
This one is unnecessary, as the recursive call will already do `#lazy` a few lines above, nice catch!